### DISCUSSION TITLE
Is a canvas swap chain in sRGB or linear space?

### DISCUSSION THREAD

**Discussion by magcius** [gpuweb/gpuweb#2537] [Q&A] [answered]
Feel free to point me to relevant prior art / ongoing discussion if necessary.

## Context

There are two kinds of textures and render targets: sRGB and linear. sRGB is effectively a curve encoding for linear colors, which helps better hide perceptual differences.

The typical way this works in a graphics pipeline: the GPU hardware has special support to encode from linear to this curve, and decode from this curve back to linear.

Encoding happens when selecting an sRGB format render target: when the shader code writes linear colors, GPU hardware encodes it from linear quantities into the sRGB curve, and writes the encoded sRGB value into the render target.

Decoding happens when selecting an sRGB format texture: when the shader code samples an sRGB texture, GPU hardware decodes it from the sRGB sample values into linear quantiites, and returns the decoded linear value to the shader.

Note that it is traditional for most display systems to use curve-encoded quantities. This is because it is the format understood by most display hardware (classically CRTs, but now most televisions and monitors expect curve-encoded quantities), so anything outputting directly to the screen needs to use curve-encoded quantities.

For instance, the [Direct3D 11 documentation](https://docs.microsoft.com/en-us/windows/win32/direct3ddxgi/converting-data-color-space) tells us:

> The app must always place sRGB data into back buffers with integer-valued formats to present the sRGB data to the screen, even if the data doesn't have this format modifier in its format name.

One way to imagine this is that the Windows compositor *always* reads a texture with a sRGB texture format. If the app chooses to use an sRGB format for its rendering, then it gets the automatic curve-encoding behavior of the GPU. Otherwise, it will put linear quantities into what should be sRGB curve-encoded colors, which is a classic mistake (sometimes called "gamma-space rendering", or "gamma-space lighting")

Of course, browser composition pipelines are more complicated than just the Windows compositor on its own.

## So, some questions:
* When specifying bgra8unorm for a swap chain, will the browser then composite the texture as-if it was sRGB?
* Will this answer change in certain conditions (different platforms? full-screening? Other conditions to enable DWM layer compositing?)
* Is this information written down anywhere, or tested?
* Could it be possible to use a bgra8srgb render target for a swap chain to enable the curve-encoding functionality of the hardware, if desired?

**Comment by kainino0x** [comment, marked as answer]
The canvas swap chain's color space is configurable via [GPUCanvasConfiguration.colorSpace](https://gpuweb.github.io/gpuweb/#dom-gpucanvasconfiguration-colorspace). At this time, that can only be `"srgb"`, which means yes, the data in the canvas is sRGB encoded - equivalent to 2d and webgl canvas defaults. In the future it should have options like `"srgb-linear"` (intended for use with higher bit depths) and `"display-p3"`. In all cases the canvas is then supposed to be color managed, though it's browser dependent whether this actually take the physical display's color space/calibration into account.
(Note: `"srgb"` here refers to the sRGB color space, not the `-srgb` texture formats.)

> * Could it be possible to use a bgra8srgb render target for a swap chain to enable the curve-encoding functionality of the hardware, if desired?

Previously the spec said it was possible to specify both `-srgb` and non-`-srgb` texture formats in `configure()`. This was removed in #2522; in ~#2336~ #2540 it will instead be possible to freely specify an `-srgb` format in `createView` so you don't need to do that in `configure()`. One reason I wanted to do this was to remove any ambiguity about whether `bgra8unorm` vs `bgra8unorm-srgb` changes how the compositor reads the texture (it would not; it would only change how the application views it).

> * Is this information written down anywhere, or tested?

It's tested: https://github.com/gpuweb/cts/blob/main/src/webgpu/web_platform/reftests/canvas_complex.html.ts
although the number of different configurations we test is limited (but [actively improving](https://github.com/gpuweb/cts/pull/952)).

As for the spec, it's not fully specified yet, but it will be when we define exactly what `GPUCanvasConfiguration.colorSpace` means. For now we have partial info: https://gpuweb.github.io/gpuweb/#canvas-color-space
The behavior also de facto matches the rest of the platform (webgl and 2d canvas defaults) which behave the same way.

**Reply by magcius** [reply]
Thank you for the detailed reply! This helped me clear up the remaining ambiguities.

> One reason I wanted to do this was to remove any ambiguity about whether bgra8unorm vs bgra8unorm-srgb changes how the compositor reads the texture (it would not; it would only change how the application views it).

Isn't this resolution the opposite of what might be wanted? If the texture is always interpreted as sRGB, wouldn't we want the user to specify a srgb texture by default, and then give them the option to reinterpret as unorm if possible?

I've also noticed some of the performance concerns around format reinterpretation as being a blocker for putting that into the spec, but wouldn't this mean that these are *always* applying on those platforms, since the platform is always reinterpreting the unorm data as sRGB on the composite side?

**Reply by kainino0x** [reply]
> Isn't this resolution the opposite of what might be wanted? If the texture is always interpreted as sRGB, wouldn't we want the user to specify a srgb texture by default, and then give them the option to reinterpret as unorm if possible?

This is a hard question to answer (that I don't fully understand the answer to), but in short: no, composition, at least inside the browser, does not actually occur in linear space, for historical reasons, despite the fact that compositing in non-linear space is mathematically incorrect. See: https://erikmcclure.com/blog/everyone-does-srgb-wrong-because/
Even worse, at least on my Mac (in all three browsers), composition doesn't even occur consistently in sRGB-non-linear space. Instead, I think it happens after color correction from sRGB to the display's calibrated space. In [this codepen](https://codepen.io/kainino0x/pen/YzEPEMo), for example, I get different results on all three of my monitors: on my side monitor, the two canvases are exactly the same color, on my main monitor they're very different, and on my laptop's internal display they're VERY different.

> I've also noticed some of the performance concerns around format reinterpretation as being a blocker for putting that into the spec, but wouldn't this mean that these are _always_ applying on those platforms, since the platform is always reinterpreting the unorm data as sRGB on the composite side?

So, I don't think the platform generally actually reinterprets the data as srgb. If a system compositor trying to do linear blending (say, between a translucent window and its background - not sure if any platform does this), then I would expect they might do it another way, like in a shader.

We're aware of at least one performance issue with srgb, which is that Intel Gen9 can't do lossless texture compression on srgb formats. (Which again hints that the compositor probably doesn't use it as srgb.) However I could imagine there are platforms where _any_ kind of reinterpretation at all causes deoptimization. And under numerous circumstances, the driver actually has no idea what kind of reinterpretation we are going to do (see below), so even just srgb reinterpretation forces us into that bucket sometimes.

I needed to go back and refresh myself on [how D3D works](https://github.com/gpuweb/gpuweb/issues/744#issuecomment-639849015). Even if assuming `CastingFullyTypedFormatSupported` (~[required since Win10 RS2](https://microsoft.github.io/DirectX-Specs/d3d/RelaxedCasting.html)~), Damyan says "CastingFullyTypedFormatSupported does not give any promises about performance" - ~though since the driver doesn't know casting will occur ahead of time, I guess on, say, Intel Gen9, the driver has to do some dynamic changes under the hood as soon as it sees a cast occur, or otherwise suffer disabling lossless texture compression completely. (I'm going to have to ask Intel/Microsoft about this.)~

On Vulkan, we can use `VK_KHR_image_format_list` when present, but otherwise are forced to use the overbroad `VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT`. I'm beginning to wonder if we need to allow srgb formats in `configure()` so users aren't forced to use `viewFormats`.

**Reply by kainino0x** [reply]
Rafael confirms it _is_ possible to use newer Windows >= 1703 with older drivers without `CastingFullyTypedFormatSupported` available, so I suspect Gen9 would fall into that category.

**Comment by kdashg** [comment]
Hold up, "srgb" here is *not* an srgb texture format, and is therefore *not* srgb-encoded. (Can you see why I generally have been objecting to these names? :))

I believe this is still largely correct: https://hackmd.io/qkU7CRnkR1iKBDQwNj570w

"srgb" in this case means srgb-gamut with perceptually-linear gamma curve. (0.5 "feels" like half way between the brightness of 0.0 and 1.0, but it's only ~21% of the photons)
"srgb-linear" is also srgb-gamut, but physically-linear, so gamma=1.0. (0.5 = 50% of the photons, even though this looks perceptually like ~73% brightness)

srgb *formats* are this other thing, where 0.5 encodes as 0xbc intead of 0x80. The reason you want to use a srgb format is because if you encode to physical-/photon-uniformly unorm8, you get low perceptual resolution in lows (blacks) and wasted perceptual resolution in highs (whites). So you use "srgb formats" when you're doing (physically-)linear color work to preserve good perceptual gradients.

So the happy paths are:
1. colorSpace: "srgb", format: "rgba8unorm"
2. colorSpace: "srgb-linear", format: "rgba8unorm-srgb" (adds photon counts, buckets/quantizes perceptually)

This space ends up as a layer-cake of cancelling curves.
If you're in #2 here, you can just bitcast to #1 and get the same result. I don't recommend treating this as a first-class equivalence though, because conceptually they're different, and the differences are important.

But that's the story of how GL called perceptual "linear" and linear "srgb".
"Linear to what?" should be your ever-present question!

**Reply by kainino0x** [reply]
As always, while I don't completely align with your phrasing, this is correct (in particular the two happy paths). I don't think it's overall contradictory with the question or my answer, though I agree that the phrasing in the question "There are two kinds of textures and render targets: sRGB and linear" isn't precise enough. Really one performs encoding/decoding and one does not.

To me "srgb" in a texture format means "the underlying bytes are gamma-encoded; when you read/write through this texture you get gamma-unencoded values" - though I avoid the word "sRGB" in this definition because srgb formats have nothing to do with color spaces (everything is just numbers, there are no colors) except that they implement one particular aspect of color encodings which use this particular gamma curve (namely, sRGB and Display P3).

**Reply by magcius** [reply]
Right, it's partly poor phrasing on my part, and of course partly legacy choices from graphics APIs (e.g. D3D11 has SRGB formats and UNORM targets. What the SRGB on the format tells you is that the GPU will automatically decode from SRGB encoding to linear during sampling, and encode from linear during pixel export).

I don't know of a better phrasing for suggesting "this pixel is encoded on a gamma curve" other than "sRGB curve-encoded", which is the language I used in my question. Perhaps something OETF/EOTF?
